libcfnet/client_code.c

```diff
@@ -869,40 +859,48 @@ bool CopyRegularFileNet(const char *source, const char *dest, off_t size,
             Log(LOG_LEVEL_INFO, "Source '%s:%s' changed while copying",
                 conn->this_server, source);
             close(dd);
-            free(buf);
             return false;
         }
 
         /* Check for mismatch between encryption here and on server. */
 
         int value = -1;
-        sscanf(buf, "t %d", &value);
+        rc = sscanf(buf, "t %d", &value);
+        if (rc != 1)
+        {
+            value = -1;
+        }
 
         if ((value > 0) && (strncmp(buf + CF_INBAND_OFFSET, "BAD: ", 5) == 0))
         {
             Log(LOG_LEVEL_INFO, "Network access to cleartext '%s:%s' denied",
                 conn->this_server, source);
             close(dd);
-            free(buf);
             return false;
         }
 
         bool w_ok = FileSparseWrite(dd, buf, n_read,
                                     &last_write_made_hole);
-        if (!w_ok)
+        n_wrote_total += n_read;
+        if (!w_ok || n_wrote_total > size)
         {
-            Log(LOG_LEVEL_ERR,
-                "Local disk write failed copying '%s:%s' to '%s'",
-                conn->this_server, source, dest);
-            free(buf);
+            if (!w_ok)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Local disk write failed copying '%s:%s' to '%s'",
+                    conn->this_server, source, dest);
+            }
+            else // if (n_wrote_total > size)
+            {
+                Log(LOG_LEVEL_INFO, "Source '%s:%s' changed while copying",
+                    conn->this_server, source);
+            }
             unlink(dest);
             close(dd);
-            FlushFileStream(conn->conn_info->sd, size - n_wrote_total - n_read);
+            FlushFileStream(conn->conn_info->sd, size - n_wrote_total);
```

Comment:
Doesn't this depend on which error case we are handling? I'd probably just check `n_wrote_total + n_read` in the condition instead of doing the addition first.